No matching begin() and end() for some infinite ranges

This code

```
#include <iostream>
#include <range/v3/view.hpp>

int main() {
    namespace rng = ranges::v3;
    namespace view = ranges::v3::view;

    auto rng1 = view::iota(1) | view::chunk(6) | view::take(3);
    for (auto&& o1 : rng1) {
        auto rng2 = o1 | view::drop(1);

        for (auto&& o2 : rng2) {
            std::cout << o2 << std::endl;
        }
    }
}
```
produces the following error on both GCC 7.2.0 and Clang 5.0.0 when using the range-v3 master branch (df4ea7b5a6dfcc1c6143045e31052aa6e2a7e56f):

> error: no matching function for call to object of type 'const _begin_::fn'

If I replace `view::iota(1)` with e.g. `view::iota(1, 30)` the code will compile just fine (hence the title of this issue).

The only compiler switches I've used are  `-std=c++17` and `-I` for including range-v3.

This code also compiles fine (with some warnings) if I use range-v3 on commit c498d86131a37a6e7d8feecb7715a04a4a63f793.

The full Clang error message is:

```
range/v3/view/drop.hpp:56:29: error: no matching function for call to object of type 'const _begin_::fn'
                return next(ranges::begin(rng_), n_, ranges::end(rng_));
                            ^~~~~~~~~~~~~
range/v3/view/drop.hpp:77:30: note: in instantiation of member function
      'ranges::v3::drop_view<ranges::v3::take_view<ranges::v3::iterator_range<ranges::v3::_basic_iterator_::basic_iterator<ranges::v3::iota_view<int, void> >, ranges::v3::default_sentinel> > >::get_begin_' requested here
                return this->get_begin_(RandomAccessRange<Rng>{});
                             ^
test.cpp:15:24: note: in instantiation of member function 'ranges::v3::drop_view<ranges::v3::take_view<ranges::v3::iterator_range<ranges::v3::_basic_iterator_::basic_iterator<ranges::v3::iota_view<int, void> >,
      ranges::v3::default_sentinel> > >::begin' requested here
        for (auto&& o2 : rng2) {
                       ^
range/v3/begin_end.hpp:75:32: note: candidate template ignored: substitution failure [with R = const
      ranges::v3::take_view<ranges::v3::iterator_range<ranges::v3::_basic_iterator_::basic_iterator<ranges::v3::iota_view<int, void> >, ranges::v3::default_sentinel> >]: no matching function for call to 'impl_'
                constexpr auto operator()(R &r) const
                               ^
range/v3/begin_end.hpp:83:32: note: candidate template ignored: substitution failure [with R =
      ranges::v3::take_view<ranges::v3::iterator_range<ranges::v3::_basic_iterator_::basic_iterator<ranges::v3::iota_view<int, void> >, ranges::v3::default_sentinel> >]: no matching function for call to 'impl_'
                constexpr auto operator()(const R &&r) const
                               ^
range/v3/begin_end.hpp:90:32: note: candidate template ignored: could not match 'reference_wrapper' against 'take_view'
                constexpr auto operator()(std::reference_wrapper<T> ref) const
                               ^
range/v3/begin_end.hpp:97:32: note: candidate template ignored: could not match 'reference_wrapper' against 'take_view'
                constexpr auto operator()(ranges::reference_wrapper<T> ref) const

```

--- a/include/range/v3/view/drop.hpp
+++ b/include/range/v3/view/drop.hpp
@@ -16,12 +16,12 @@
 
 #include <type_traits>
 #include <meta/meta.hpp>
-#include <range/v3/detail/satisfy_boost_range.hpp>
+#include <range/v3/iterator_range.hpp>
+#include <range/v3/range_concepts.hpp>
 #include <range/v3/range_fwd.hpp>
 #include <range/v3/range_traits.hpp>
-#include <range/v3/range_concepts.hpp>
 #include <range/v3/view_interface.hpp>
-#include <range/v3/iterator_range.hpp>
+#include <range/v3/detail/satisfy_boost_range.hpp>
 #include <range/v3/utility/box.hpp>
 #include <range/v3/utility/functional.hpp>
 #include <range/v3/utility/iterator_traits.hpp>
@@ -50,14 +50,20 @@ namespace ranges
             Rng rng_;
             difference_type_ n_;
 
-            // RandomAccessRange == true
-            iterator_t<Rng> get_begin_(std::true_type) const
+            template<typename BaseRng = Rng>
+            iterator_t<BaseRng const> get_begin_(std::true_type, std::true_type) const
             {
+                CONCEPT_ASSERT(RandomAccessRange<Rng const>());
                 return next(ranges::begin(rng_), n_, ranges::end(rng_));
             }
-            // RandomAccessRange == false
-            iterator_t<Rng> get_begin_(std::false_type)
+            iterator_t<Rng> get_begin_(std::true_type, std::false_type)
             {
+                CONCEPT_ASSERT(RandomAccessRange<Rng>());
+                return next(ranges::begin(rng_), n_, ranges::end(rng_));
+            }
+            iterator_t<Rng> get_begin_(std::false_type, detail::any)
+            {
+                CONCEPT_ASSERT(!RandomAccessRange<Rng>());
                 using cache_t = detail::non_propagating_cache<
                     iterator_t<Rng>, drop_view<Rng>>;
                 auto &begin_ = static_cast<cache_t&>(*this);
@@ -72,10 +78,12 @@ namespace ranges
             {
                 RANGES_EXPECT(n >= 0);
             }
+            CONCEPT_REQUIRES(!RandomAccessRange<Rng const>())
             iterator_t<Rng> begin()
             {
-                return this->get_begin_(RandomAccessRange<Rng>{});
+                return this->get_begin_(RandomAccessRange<Rng>{}, std::false_type{});
             }
+            CONCEPT_REQUIRES(!RandomAccessRange<Rng const>())
             sentinel_t<Rng> end()
             {
                 return ranges::end(rng_);
@@ -84,7 +92,7 @@ namespace ranges
                 CONCEPT_REQUIRES_(RandomAccessRange<BaseRng const>())>
             iterator_t<BaseRng const> begin() const
             {
-                return this->get_begin_(std::true_type{});
+                return this->get_begin_(std::true_type{}, std::true_type{});
             }
             template<typename BaseRng = Rng,
                 CONCEPT_REQUIRES_(RandomAccessRange<BaseRng const>())>
@@ -99,7 +107,7 @@ namespace ranges
                 auto const n = static_cast<range_size_type_t<Rng>>(n_);
                 return s < n ? 0 : s - n;
             }
-            CONCEPT_REQUIRES(SizedRange<Rng>())
+            CONCEPT_REQUIRES(!SizedRange<Rng const>() && SizedRange<Rng>())
             range_size_type_t<Rng> size()
             {
                 auto const s = static_cast<range_size_type_t<Rng>>(ranges::size(rng_));

--- a/test/view/drop.cpp
+++ b/test/view/drop.cpp
@@ -120,5 +120,17 @@ int main()
         ::check_equal(rng, {2,3});
     }
 
+    {
+        // regression test for #728
+        auto rng1 = view::iota(1) | view::chunk(6) | view::take(3);
+        int i = 2;
+        RANGES_FOR(auto o1, rng1)
+        {
+            auto rng2 = o1 | view::drop(1);
+            ::check_equal(rng2, {i, i+1, i+2, i+3, i+4});
+            i += 6;
+        }
+    }
+
     return test_result();
 }

